app/gui2/src/util/ast/abstract.ts

```diff
@@ -126,6 +139,86 @@ export function deleteFromParentBlock(ast: MutableAst) {
     parent.updateLines((lines) => lines.filter((line) => line.expression?.node.id !== ast.id))
 }
 
+/** If the input is a chain of applications of the given left-associative operator, and all the leaves of the
+ *  operator-application tree are identifier expressions, return the identifiers from left to right.
+ *  This is analogous to `ast.code().split(operator)`, but type-enforcing.
+ */
+export function unrollOprChain(
+  ast: Ast,
+  leftAssociativeOperator: string,
+): IdentifierOrOperatorIdentifier[] | null {
+  const idents: IdentifierOrOperatorIdentifier[] = []
+  let ast_: Ast | undefined = ast
+  while (
+    ast_ instanceof OprApp &&
+    ast_.operator.ok &&
+    ast_.operator.value.code() === leftAssociativeOperator
+  ) {
+    if (!(ast_.rhs instanceof Ident)) return null
+    idents.unshift(ast_.rhs.code())
+    ast_ = ast_.lhs
+  }
+  if (!(ast_ instanceof Ident)) return null
+  idents.unshift(ast_.code())
+  return idents
+}
+
+/** If the input is a chain of property accesses (uses of the `.` operator with a syntactic identifier on the RHS), and
+ *  the value at the beginning of the sequence is an identifier expression, return all the identifiers from left to
+ *  right. This is analogous to `ast.code().split('.')`, but type-enforcing.
+ */
+export function unrollPropertyAccess(ast: Ast): IdentifierOrOperatorIdentifier[] | null {
+  const idents: IdentifierOrOperatorIdentifier[] = []
+  let ast_: Ast | undefined = ast
+  while (ast_ instanceof PropertyAccess) {
+    idents.unshift(ast_.rhs.code())
+    ast_ = ast_.lhs
+  }
+  if (!(ast_ instanceof Ident)) return null
+  idents.unshift(ast_.code())
+  return idents
+}
+
+export function parseIdent(ast: Ast): IdentifierOrOperatorIdentifier | null {
+  if (ast instanceof Ident) {
+    return ast.code()
+  } else {
+    return null
+  }
+}
+
+export function parseIdents(ast: Ast): IdentifierOrOperatorIdentifier[] | null {
+  return unrollOprChain(ast, ',')
+}
+
+export function parseQualifiedName(ast: Ast): QualifiedName | null {
+  const idents = unrollPropertyAccess(ast)
+  return idents && normalizeQualifiedName(qnFromSegments(idents))
+}
+
+/* Substitute `name` inside `expression` with `to`. */
+export function substituteQualifiedName(
```

Comment:
There is a major flaw in the current implementation, and I’m not sure how to overcome it.
The failing test case would be:
```ts
// …
const expr = Ast.parse('Name', edit)
substituteQualifiedName(edit, expr, 'Name', 'OtherName' as QualifiedName)
```
This will *not* change `expr`, because `updateValue` requires some parent, but `expr` has none.

I overcome it by doing conflict resolution later, when we already have assignment ast: `variable = expr`, and in this case `expr` always has the parent, so `updateValue` works.

How it should be done?
cc @kazcw